Fix deprecated code snippets and removed pre-6.3 code snippets (#233)
smartdevicelink/sdl_app_library_guides: docs/Other SDL Features/Calling a Phone Number/index.md

# Calling a Phone Number
The @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ RPC allows you make a phone call via the user's phone. In order to make dial a phone number you must be sure that the device is connected via Bluetooth (even if using USB) for this request to work. If the phone is not connected via Bluetooth, you will receive a result of `REJECTED` from the module.

## Checking Your App's Permissions
@![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ is an RPC that is usually restricted by OEMs. As a result, a module may reject your request if your app does not have the correct permissions. Your SDL app may also be restricted to only being allowed to making a phone call when your app is open (i.e. the `hmiLevel` is non-`NONE`) or when it is the currently active app (i.e. the `hmiLevel` is `FULL`). 

@![iOS]
##### Objective-C
```objc
// TODO
```

##### Swift
```swift

> _  = sdlManager.permissionManager.addObserver(forRPCs: [SDLRPCFunctionName.dialNumber.rawValue.rawValue], groupType:.any, withHandler: { (individualStatuses, groupStatus) in

Reviewer: ```suggestion
let observerId  = sdlManager.permissionManager.addObserver(forRPCs: [SDLRPCFunctionName.dialNumber.rawValue.rawValue], groupType: .any, withHandler: { (allChanges, groupStatus) in
```

Reviewer: We should add an API to take the `SDLRPCFunctionName` without using `rawValue` now that we have it.

>     guard groupStatus == SDLPermissionGroupStatus.allowed else { 

Reviewer: ```suggestion
    // This handler will be called whenever the permission status changes
    guard groupStatus == SDLPermissionGroupStatus.allowed else { 
```

        // Your app does not have permission to send the `DialNumber` request for its current HMI level
        return 
    }

    // Your app has permission to send the `DialNumber` request for its current HMI level
})
```
!@

@![android,javaSE,javaEE]
```java
// TODO
```
!@

## Checking if the Module Supports Calling a Phone Number
Since making a phone call is a newer feature, there is a possibility that some legacy modules will not support your request because the module does not support the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request. Once you have successfully connected to the module, you can check the module's capabilities via the @![iOS]`SDLManager.systemCapabilityManager`!@@![android, javaSE, javaEE]`sdlManager.getSystemCapabilityManager`!@ as shown below:

@![iOS]
##### Objective-C
```objc
// TODO
```

##### Swift
```swift

> // Check if the module supports making a phone call

Reviewer: ```suggestion
// Check if the module supports has phone capabilities
```

> guard (sdlManager.systemCapabilityManager.isCapabilitySupported(type: .phoneCall)) else { return }

Reviewer: Turn this into a method to check if `DialNumber` is supported, or something like that. You don't want to call `updateCapabilityType` _every_ time you are going to go to call `DialNumber` so the value should probably be cached. Maybe make a small class to handle checking and sending `DialNumber`? Or a method? Also, tell the dev to disable any buttons that would send `DialNumber` if not supported / the app doesn't have permissions.

Author: Added a method and a paragraph telling devs to disable the buttons

// Check if the module supports the `DialNumber` request
sdlManager.systemCapabilityManager.updateCapabilityType(.phoneCall) { [weak self] (error, systemCapabilityManager) in
    let isDialNumberSupported: Bool
    if let phoneCapability = systemCapabilityManager.phoneCapability {
        isDialNumberSupported = phoneCapability.dialNumberEnabled?.boolValue ?? false
    } else {
        // Legacy modules (pre RPC Spec v4.5) do not support the phone call capability so for versions less than 4.5 we will assume `DialNumber` is supported
        let rpcSpecVersion = self?.sdlManager.registerResponse?.sdlMsgVersion ?? SDLMsgVersion(majorVersion: 0, minorVersion: 0, patchVersion: 0)

>         isDialNumberSupported = "\(rpcSpecVersion.majorVersion).\(rpcSpecVersion.minorVersion)".compare("4.5", options: .numeric) == .orderedAscending

Reviewer: This would be done _before_ calling `updateCapabilityType`

Author: Fixed

>         // Legacy modules (pre RPC Spec v4.5) do not support the phone call capability so for versions less than 4.5 we will assume `DialNumber` is supported
>         let rpcSpecVersion = self?.sdlManager.registerResponse?.sdlMsgVersion ?? SDLMsgVersion(majorVersion: 0, minorVersion: 0, patchVersion: 0)
>         isDialNumberSupported = "\(rpcSpecVersion.majorVersion).\(rpcSpecVersion.minorVersion)".compare("4.5", options: .numeric) == .orderedAscending

Reviewer: ```suggestion
        // Legacy modules (pre RPC Spec v4.5) do not support the phone call capability so for versions less than 4.5 we will assume `DialNumber` is supported
        let rpcSpecVersion = self?.sdlManager.registerResponse?.sdlMsgVersion ?? SDLMsgVersion(majorVersion: 0, minorVersion: 0, patchVersion: 0)
        isDialNumberSupported = "\(rpcSpecVersion.majorVersion).\(rpcSpecVersion.minorVersion)".compare("4.5", options: .numeric) == .orderedAscending
```

Instead, convert to `SDLVersion` and use that to compare.

Author: Fixed

# Calling a Phone Number

> The @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ RPC allows you make a phone call via the user's phone. In order to make dial a phone number you must be sure that the device is connected via Bluetooth (even if using a USB cord) for this request to work. If the phone is not connected via Bluetooth, you will receive a result of `REJECTED` from the module.

Reviewer: ```suggestion
The @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ RPC allows you make a phone call via the user's phone. In order to dial a phone number you must be sure that the device is connected via Bluetooth (even if using a USB cord) for this request to work. If the phone is not connected via Bluetooth, you will receive a result of `REJECTED` from the module.
```

## Checking Your App's Permissions
@![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ is an RPC that is usually restricted by OEMs. As a result, a module may reject your request if your app does not have the correct permissions. Your SDL app may also be restricted to only being allowed to making a phone call when your app is open (i.e. the `hmiLevel` is non-`NONE`) or when it is the currently active app (i.e. the `hmiLevel` is `FULL`). 

@![iOS]
##### Objective-C
```objc
// TODO
```

##### Swift
```swift

> let observerId  = sdlManager.permissionManager.addObserver(forRPCs: [SDLRPCFunctionName.dialNumber.rawValue.rawValue], groupType: .any, withHandler: { (allChanges, groupStatus) in

Reviewer: ```suggestion
let observerId = sdlManager.permissionManager.addObserver(forRPCs: [SDLRPCFunctionName.dialNumber.rawValue.rawValue], groupType: .any, withHandler: { (allChanges, groupStatus) in
```

    // This handler will be called whenever the permission status changes
    guard groupStatus == SDLPermissionGroupStatus.allowed else { 
        // Your app does not have permission to send the `DialNumber` request for its current HMI level
        return 
    }

    // Your app has permission to send the `DialNumber` request for its current HMI level
})
```
!@

@![android,javaSE,javaEE]
```java
// TODO
```
!@

## Checking if the Module Supports Calling a Phone Number

> Since making a phone call is a newer feature, there is a possibility that some legacy modules will reject your request because the module does not support the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request. Once you have successfully connected to the module and you know that your SDL app has been opened, you can check the module's capabilities via the @![iOS]`SDLManager.systemCapabilityManager`!@@![android, javaSE, javaEE]`sdlManager.getSystemCapabilityManager`!@ as shown in the example below. Please note that you only need to check once if the module supports calling a phone number, however you must wait to perform this check until you know that the SDL app has been opened (i.e. the `hmiLevel` is non-`NONE`). 

Reviewer: ```suggestion
Since making a phone call is a newer feature, there is a possibility that some legacy modules will reject your request because the module does not support the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request. Once you have successfully connected to the module, you can check the module's capabilities via the @![iOS]`SDLManager.systemCapabilityManager`!@@![android, javaSE, javaEE]`sdlManager.getSystemCapabilityManager`!@ as shown in the example below. Please note that you only need to check once if the module supports calling a phone number, however you must wait to perform this check until you know that the SDL app has been opened (i.e. the `hmiLevel` is non-`NONE`). 
```

> If you discover that the module does not support calling a phone number or that your app does not have the right permissions, you should disable any buttons, voice commands, menu items, etc. in your app that would send the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request.

Reviewer: ```suggestion
!!! NOTE
If you discover that the module does not support calling a phone number or that your app does not have the right permissions, you should disable any buttons, voice commands, menu items, etc. in your app that would send the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request.
!!!
```

@![iOS]
##### Objective-C
```objc
// TODO
```

##### Swift
```swift

> func isDialNumberSupported(handler: @escaping (_ success: Bool, _ error: Error?) -> ()) {

Reviewer: ```suggestion
func isDialNumberSupported(handler: @escaping (_ success: Bool, _ error: Error?) -> Void) {
```

    // Check if the module has phone capabilities
    guard (sdlManager.systemCapabilityManager.isCapabilitySupported(type: .phoneCall)) else {
        return handler(false, nil)
    }

>     // Legacy modules (pre RPC Spec v4.5) do not support the phone call capability so for versions less than 4.5 we will assume `DialNumber` is supported

Reviewer: ```suggestion
    // Legacy modules (pre-RPC Spec v4.5) do not support checking capabilities, so for versions less than 4.5 we will assume `DialNumber` is supported
```

    guard let sdlMsgVersion = sdlManager.registerResponse?.sdlMsgVersion, SDLVersion(sdlMsgVersion: sdlMsgVersion).isGreaterThanOrEqual(to: SDLVersion(major: 4, minor: 5, patch: 0)) else {
        return handler(true, nil)
    }

    // Check if the module supports the `DialNumber` request

>     sdlManager.systemCapabilityManager.updateCapabilityType(.phoneCall) { (error, systemCapabilityManager) in

Reviewer: This should check the cached `phoneCapability` before calling `updateCapabilityType`

Author: Added.

Reviewer: It doesn't appear that it was

Author: Sorry. I forgot to push. It should be fixed now.

        if (error != nil) {
            return handler(false, error)
        }

>         let isDialNumberSupported: Bool
>         if let phoneCapability = systemCapabilityManager.phoneCapability {
>             isDialNumberSupported = phoneCapability.dialNumberEnabled?.boolValue ?? false
>         } else {
>             isDialNumberSupported = false
>         }

Reviewer: ```suggestion
        let isDialNumberSupported = false
        if let phoneCapability = systemCapabilityManager.phoneCapability {
            isDialNumberSupported = phoneCapability.dialNumberEnabled?.boolValue ?? false
        }
```

        return handler(isDialNumberSupported, nil)
    }
}
```
!@

@![android,javaSE,javaEE]
```java
// TODO
```
!@

## Sending a Dial Number Request

> Once you know that the module supports dialing a phone number and that your SDL app has permission to send a @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request, you can create and send the request. 

Reviewer: ```suggestion
Once you know that the module supports dialing a phone number and that your SDL app has permission to send the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request, you can create and send the request. 
```

!!! NOTE
If you discover that the module does not support calling a phone number or that your app does not have the right permissions, you should disable any buttons, voice commands, menu items, etc. in your app that would send the @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ request.
!!!

@![iOS]
##### Objective-C
```objc
// TODO
```

##### Swift
```swift
func isDialNumberSupported(handler: @escaping (_ success: Bool, _ error: Error?) -> Void) {
    // Check if the module has phone capabilities
    guard (sdlManager.systemCapabilityManager.isCapabilitySupported(type: .phoneCall)) else {
        return handler(false, nil)
    }

>     // Legacy modules (pre-RPC Spec v4.5) do not support checking capabilities, so for versions less than 4.5 we will assume `DialNumber` is supported

Reviewer: ```suggestion
    // Legacy modules (pre-RPC Spec v4.5) do not support system capabilities, so for versions less than 4.5 we will assume `DialNumber` is supported
```

!!!

## Checking if Dial Number is Available

> @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ is a newer RPC, so there is a possibility that not all head units will support it. Once you have successfully connecting to the module, you can access the phone capability property on the @![iOS]`SDLManager.systemCapabilityManager`!@@![android, javaSE, javaEE]`sdlManager.getSystemCapabilityManager`!@ instance.

Author: When you say **full property access** do you mean like this: `SDLManager.systemCapabilityManager.subscribe(capabilityType:updateHandler:)` ? I was trying to keep it simple since the code sample is below.

Not sure what needs to be checked using the Permission Manger.

> @![iOS]`SDLDialNumber`!@@![android,javaSE,javaEE]`DialNumber`!@ is a newer RPC, so there is a possibility that not all head units will support it. Once you have successfully connecting to the module, you can access the phone capability property on the @![iOS]`SDLManager.systemCapabilityManager`!@@![android, javaSE, javaEE]`sdlManager.getSystemCapabilityManager`!@ instance.

Reviewer: Sorry, this was from an earlier comment before I understood what you were doing.

However, I have a few other notes:
1. Before calling `subscribe`, the dev ought to call `isCapabilitySupported`.
2. The permission manager part would be to check if the dev actually _has permission_ to use `DialNumber`. I thought that it could also be used to check if `DialNumber` is supported to begin with. Your current method will work on some HUs, but between v3.0 (when `DialNumber` was introduced), and v4.5 (when `GetSystemCapability` was introduced), it will not.

@![android,javaSE,javaEE]
```java

> sdlManager.getSystemCapabilityManager().getCapability(SystemCapabilityType.PHONE_CALL, new OnSystemCapabilityListener() {

Author: I updated the code snippet. Not sure what you mean about `getCapability()` not working since there is a fallback check for `SystemCapabilityType.PHONE_CALL`.

> sdlManager.getSystemCapabilityManager().getCapability(SystemCapabilityType.PHONE_CALL, new OnSystemCapabilityListener() {

Reviewer: See above comment, does that explain better?

	@Override
	public void onCapabilityRetrieved(Object capability) {
		PhoneCapability phoneCapability = (PhoneCapability) capability;
		Boolean isDialNumberSupported = ((PhoneCapability) capability).getDialNumberEnabled().booleanValue();
		// If making phone calls is supported, send the `DialNumber` RPC
	}

	@Override
	public void onError(String info) {
		// Perform a fallback check because the module does not support the phone capability
		isPhoneCallAvailable();
	}
}, false);

> private void isPhoneCallAvailable() {

Author: Fixed

        return handler(true, nil)
    }

    // Check if the phone capability has already been retrieved from the module
    if let phoneCapability = sdlManager.systemCapabilityManager.phoneCapability {

>         // Check if the module supports the `DialNumber` request

Reviewer: I don't believe that this is the correct comment here.

Author: Could you clarify what the correct comment would be?
